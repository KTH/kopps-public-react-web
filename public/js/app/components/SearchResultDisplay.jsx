import React, { useEffect, useState } from 'react'

import axios from 'axios'
import ReactDOM from 'react-dom'
import { useHistory } from 'react-router-dom'
import PropTypes from 'prop-types'

import { useStore } from '../mobx'
import SearchTableView, { searchHitsPropsShape } from './SearchTableView'

import i18n from '../../../../i18n'
import { stringifyUrlParams, CLIENT_EDU_LEVELS, CLIENT_SHOW_OPTIONS } from '../../../../domain/searchParams'
// eslint-disable-next-line import/no-cycle
import { SearchAlert } from './index'
import { STATUS, ERROR_ASYNC, useAsync } from '../hooks/searchUseAsync'

function _getThisHost(thisHostBaseUrl) {
  return thisHostBaseUrl.slice(-1) === '/' ? thisHostBaseUrl.slice(0, -1) : thisHostBaseUrl
}

// eslint-disable-next-line consistent-return
async function koppsCourseSearch(language, proxyUrl, params) {
  try {
    const result = await axios.get(`${proxyUrl}/intern-api/sok/${language}`, {
      params,
    })
    if (result) {
      if (result.status >= 400) {
        return 'ERROR-koppsCourseSearch-' + result.status
      }
      const { data } = result
      return data
    }
  } catch (error) {
    if (error.response) {
      throw new Error('Unexpected error from koppsCourseSearch-' + error.message)
    }
    throw error
  }
}

function renderAlertToTop(errorType, languageIndex) {
  const alertContainer = document.getElementById('alert-placeholder')
  if (alertContainer) {
    ReactDOM.render(<SearchAlert alertType={errorType} languageIndex={languageIndex} />, alertContainer)
  }
}
function dismountTopAlert() {
  const alertContainer = document.getElementById('alert-placeholder')
  if (alertContainer) ReactDOM.unmountComponentAtNode(alertContainer)
}

const errorItalicParagraph = (errorType, languageIndex) => {
  const errorText = i18n.messages[languageIndex].generalSearch[errorType]
  if (!errorText)
    throw new Error(
      `Missing translations for errorType: ${errorType}. Allowed types: ${Object.values(ERROR_ASYNC).join(', ')}`
    )

  return (
    <p>
      <i>{errorText}</i>
    </p>
  )
}

function DisplayResult({ languageIndex, status, errorType, searchResults }) {
  if (status === STATUS.resolved) {
    return <SearchTableView unsortedSearchResults={searchResults} />
  }

  if (status === STATUS.idle) return null
  if (status === STATUS.pending) {
    const { searchLoading } = i18n.messages[languageIndex].generalSearch
    return <p>{searchLoading}</p>
  }
  if (errorType) return errorItalicParagraph(errorType, languageIndex)

  return null
}

DisplayResult.propTypes = {
  languageIndex: PropTypes.oneOf([0, 1]),
  status: PropTypes.oneOf([...Object.values(STATUS), null]),
  errorType: PropTypes.oneOf([...Object.values(ERROR_ASYNC), '']),
  searchResults: PropTypes.shape(searchHitsPropsShape),
}

DisplayResult.defaultProps = {
  languageIndex: 0,
  errorType: '',
  searchResults: {},
  status: null,
}

function SearchResultDisplay({ searchParameters, onlyPattern = false }) {
  const { browserConfig, language, languageIndex } = useStore()
  const history = useHistory()
  const { pattern } = searchParameters
  const searchStr = stringifyUrlParams(searchParameters)
  const [loadStatus, setLoadStatus] = useState('firstLoad')
  const { resultsHeading } = i18n.messages[languageIndex].generalSearch

  const asyncCallback = React.useCallback(() => {
    if (onlyPattern && !pattern) return
    if (!searchStr && loadStatus === 'firstLoad') {
      setLoadStatus('afterLoad')
      return
    }
    const proxyUrl = _getThisHost(browserConfig.proxyPrefixPath.uri)
    // eslint-disable-next-line consistent-return
    return koppsCourseSearch(language, proxyUrl, searchParameters)
  }, [searchParameters])

  const initialStatus = onlyPattern
    ? { status: pattern ? STATUS.pending : STATUS.idle }
    : { status: searchStr ? STATUS.pending : STATUS.idle }

  const state = useAsync(asyncCallback, initialStatus)

  const { data: searchResults, status, error: errorType } = state

  useEffect(() => {
    if (errorType && errorType !== null) {
      renderAlertToTop(errorType, languageIndex)
    } else dismountTopAlert()
  }, [status])

  useEffect(() => {
    if ((onlyPattern && pattern) || !onlyPattern) {
      if (status === STATUS.pending) {
        history.push({ search: searchStr })
      }
    }
  }, [status])

  return (
    <>
      {status !== STATUS.idle && <h2 id="results-heading">{resultsHeading}</h2>}
      <DisplayResult
        status={status}
        errorType={errorType}
        languageIndex={languageIndex}
        searchResults={searchResults}
      />
    </>
  )
}

SearchResultDisplay.propTypes = {
  onlyPattern: PropTypes.bool,
  searchParameters: PropTypes.shape({
    eduLevel: PropTypes.arrayOf(PropTypes.oneOf(CLIENT_EDU_LEVELS)),
    pattern: PropTypes.string,
    // eslint-disable-next-line consistent-return
    period: PropTypes.arrayOf((propValue, key, componentName, location, propFullName) => {
      if (!/\b\d{5}\b:(\b\d)/.test(propValue[key]) && !/\b\d{4}\b:(summer)/.test(propValue[key])) {
        return new Error(
          `Prop ${propFullName} of a component ${componentName} has incorrect prop value ${propValue[key]}`
        )
      }
    }),
    showOptions: PropTypes.arrayOf(PropTypes.oneOf(CLIENT_SHOW_OPTIONS)),
    // eslint-disable-next-line consistent-return
    department: (propValue, key, componentName, location, propFullName) => {
      if (!/[A-Za-zÅÄÖåäö]{1,4}/.test(propValue[key])) {
        return new Error(
          `Prop ${propFullName} of a component ${componentName} has incorrect prop value ${propValue[key]}`
        )
      }
    },
  }),
}

SearchResultDisplay.defaultProps = {
  onlyPattern: false,
  searchParameters: {},
}

export default SearchResultDisplay
