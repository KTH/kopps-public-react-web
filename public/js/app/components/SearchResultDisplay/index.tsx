import React, { useState } from 'react'

import './style.scss'

import { useStore } from '../../mobx'

import { SearchResultDisplayParams, CourseSearchResult, VIEW, View } from './types'
import { STATUS } from '../../hooks/types/UseCourseSearchTypes'
import Article from '../Article'
import SearchAlert from '../SearchAlert'
import SearchResultHeader from './SearchResultHeader'
import SearchResultComponent from './SearchResultComponent'
import { AlertType } from '../SearchAlert/types'

const isCourseSearchResult = (data: string | CourseSearchResult): data is CourseSearchResult => {
  return (data as CourseSearchResult).searchHits !== undefined
}
const SearchResultDisplay: React.FC<SearchResultDisplayParams> = ({ resultsState }) => {
  const { languageIndex } = useStore()
  const [view, setView] = useState<View>(VIEW.list)
  const { data: searchResults, status: searchStatus, error: errorType } = resultsState

  if (errorType) {
    return (
      <SearchAlert
        alertType={errorType}
        languageIndex={languageIndex}
      />
    )
  }

  return (
    <Article classNames={['course-search-results']}>
      <SearchResultHeader
        resultsLength={searchResults ? searchResults.searchHits.length : undefined}
        searchStatus={searchStatus}
        view={view}
        setView={setView}
      />
      {searchStatus === STATUS.resolved &&
        isCourseSearchResult(searchResults) &&
        searchResults.searchHits.length > 0 && <SearchResultComponent searchResults={searchResults} view={view} />}
    </Article>
  )
}

export default SearchResultDisplay
