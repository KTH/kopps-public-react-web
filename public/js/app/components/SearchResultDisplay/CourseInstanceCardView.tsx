import React from 'react'
import './style.scss'
import { CourseHeader } from './ListView'
import { useLanguage } from '../../hooks/useLanguage'
import { compareCourseDTOBy, createLinkElement, formatCourseInstance } from '../../util/searchHelper'
import { CourseInstanceSearchDTO } from '@kth/om-kursen-ladok-client/dist/search/types'

export const CourseInstanceCardView: React.FC<{ results: CourseInstanceSearchDTO[] }> = ({ results }) => {
  const { translation, languageShortname } = useLanguage()

  const { linkToInforKursval } = translation.generalSearch

  const { course_pace, course_campus, course_language } = translation.messages

  results.sort(compareCourseDTOBy('kod'))

  return (
    <>
      {results.map((course, index) => {
        const {
          courseCode,
          startTerm,
          title,
          credits,
          courseHasNoRounds,
          languages,
          studyPaces,
          campuses,
          periodTexts,
        } = formatCourseInstance(course, languageShortname)

        const InforKursvalLink = createLinkElement({
          text: linkToInforKursval,
          code: courseCode,
          startTerm,
          language: languageShortname,
          isExternal: true,
        })

        return (
          <div className="course-card" key={courseCode + index}>
            <CourseHeader
              title={title}
              credits={credits}
              courseCode={courseCode}
              periodTexts={periodTexts}
              courseHasNoRounds={courseHasNoRounds}
            />
            <div className="course-footer">
              <div className="course-details">
                <div className="course-location">
                  <span className="icon">{course_campus}</span>
                  {stringArrayToSpan(campuses)}
                </div>
                <div className="course-language">
                  <span className="icon">{course_language}</span>
                  {stringArrayToSpan(languages)}
                </div>
                <div className="course-pace">
                  <span className="icon">{course_pace}</span>
                  {stringArrayToSpan(studyPaces)}
                </div>
              </div>
              <div className="course-link">{InforKursvalLink}</div>
            </div>
          </div>
        )
      })}
    </>
  )
}

const stringArrayToSpan = (arr: (number | string)[]) => {
  return arr.map(str => <span key={str}>{str}</span>)
}
