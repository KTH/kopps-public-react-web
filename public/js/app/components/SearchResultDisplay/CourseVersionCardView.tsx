import React from 'react'
import './style.scss'
import { CourseVersionDTO } from '@kth/om-kursen-ladok-client/dist/search/types'
import { CourseHeader } from './ListView'
import { compareCourseDTOBy, createLinkElement } from '../../util/searchHelper'
import { useLanguage } from '../../hooks/useLanguage'

export const CourseVersionCardView: React.FC<{
  results: CourseVersionDTO[]
}> = ({ results }) => {
  results.sort(compareCourseDTOBy('kod'))

  return (
    <>
      {results.map((courseVersion: CourseVersionDTO, index: number) => (
        <CourseVersionCard key={courseVersion.kod + index} courseVersion={courseVersion} />
      ))}
    </>
  )
}

const CourseVersionCard: React.FC<{
  courseVersion: CourseVersionDTO
}> = ({ courseVersion: { kod, benamning, omfattning } }) => {
  const { translation, languageShortname } = useLanguage()

  const { linkToInforKursval } = translation.generalSearch
  const InforKursvalLink = createLinkElement({
    text: linkToInforKursval,
    code: kod,
    language: languageShortname,
    isExternal: true,
  })

  return (
    <div className="course-card">
      <CourseHeader title={benamning} credits={omfattning} courseCode={kod} />
      <div className="course-footer">
        <div className="course-details">
          <div className="course-link">{InforKursvalLink}</div>
        </div>
      </div>
    </div>
  )
}
