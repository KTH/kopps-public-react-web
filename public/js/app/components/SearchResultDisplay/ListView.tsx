import React from 'react'
import './style.scss'

import { useLanguage } from '../../hooks/useLanguage'
import { ResultType, SearchData } from '../../hooks/types/UseCourseSearchTypes'
import { CourseVersionCardView } from './CourseVersionCardView'
import { CourseInstanceCardView } from './CourseInstanceCardView'

export const ListView: React.FC<{
  searchData: SearchData
}> = ({ searchData }) => {
  const { type, results } = searchData

  switch (type) {
    case ResultType.VERSION:
      return <CourseVersionCardView results={results} />
    case ResultType.INSTANCE:
      return <CourseInstanceCardView results={results} />
  }
}

export const CourseHeader: React.FC<{
  title: string
  credits: string
  courseCode: string
  periodTexts?: string[]
  courseHasNoRounds?: boolean
}> = ({ title, credits, courseCode, periodTexts, courseHasNoRounds }) => {
  const { translation } = useLanguage()
  const { generalSearch } = translation

  return (
    <div className="course-header">
      <h3>
        {title}, {credits}
      </h3>
      <span className="course-code">{courseCode}</span>
      {periodTexts && periodTexts.length > 0 && <span className="course-period">{periodTexts.join(', ')}</span>}
      {courseHasNoRounds && <i>{generalSearch.courseHasNoRounds}</i>}
    </div>
  )
}

export default ListView
