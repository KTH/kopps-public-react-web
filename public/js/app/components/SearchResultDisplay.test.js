import React from 'react'
import { render, screen, act, waitForElementToBeRemoved } from '@testing-library/react'
import '@testing-library/jest-dom'
import koppsCourseSearch from '../util/internApi'
import SearchResultDisplay from './SearchResultDisplay'
import { useStore } from '../mobx'
import {
  TEST_API_ANSWER_ALGEBRA,
  TEST_API_ANSWER_EMPTY_PARAMS,
  TEST_API_ANSWER_OVERFLOW,
  TEST_API_ANSWER_UNKNOWN_ERROR,
  TEST_API_ANSWER_NO_HITS,
  TEST_API_ANSWER_RESOLVED,
} from './mocks/mockKoppsCourseSearch'

jest.setTimeout(1000)

jest.mock('../mobx')
jest.mock('../util/internApi')
jest.mock('react-router-dom', () => ({
  useNavigate: jest.fn(),
}))

describe.skip('Component <SearchResultDisplay> and its warning messages', () => {
  test('double search: twice render empty parameters, first time without asking kopps, second time show a warning message. English. 1A', async () => {
    useStore.mockReturnValue({
      browserConfig: { proxyPrefixPath: { uri: '/student/kurser' } },
      language: 'en',
      languageIndex: 0,
    })
    const { asFragment, rerender } = render(<SearchResultDisplay searchParameters={{}} />)

    expect(koppsCourseSearch).not.toHaveBeenCalled()

    expect(asFragment()).toMatchSnapshot()

    koppsCourseSearch.mockReturnValue(Promise.resolve(TEST_API_ANSWER_EMPTY_PARAMS))

    act(() => rerender(<SearchResultDisplay searchParameters={{}} />))
    expect(koppsCourseSearch).toHaveBeenCalledWith('en', '/student/kurser', {})
    await waitForElementToBeRemoved(() => screen.getByText(/searching/i))

    try {
      expect(screen.getByText(/Search results/i)).toBeInTheDocument()
      expect(screen.getByText(/No query restriction was specified/i)).toBeInTheDocument()
    } catch (error) {
      error.message = `${`Because it is a second try to search with empty params it must give a warning message\n\n ${error}`} `
      throw error
    }

    expect(asFragment()).toMatchSnapshot()
  })

  test('double search: by a text pattern and rerender it again with the empty parameter to get a warning message noQueryProvided. English. 2A', async () => {
    useStore.mockReturnValue({
      browserConfig: { proxyPrefixPath: { uri: '/student/kurser' } },
      language: 'en',
      languageIndex: 0,
    })

    koppsCourseSearch.mockReturnValue(Promise.resolve(TEST_API_ANSWER_ALGEBRA))

    const { asFragment, rerender } = render(<SearchResultDisplay searchParameters={{ pattern: 'Algebra' }} />)
    expect(screen.getByText(/searching/i)).toBeInTheDocument()

    expect(koppsCourseSearch).toHaveBeenCalledWith('en', '/student/kurser', { pattern: 'Algebra' })

    // rerender with empty params
    koppsCourseSearch.mockReturnValue(Promise.resolve(TEST_API_ANSWER_EMPTY_PARAMS))

    act(() => rerender(<SearchResultDisplay searchParameters={{}} />))
    expect(koppsCourseSearch).toHaveBeenCalledWith('en', '/student/kurser', {})
    await waitForElementToBeRemoved(() => screen.getByText(/searching/i))

    try {
      expect(screen.getByText(/Search results/i)).toBeInTheDocument()
      expect(screen.getByText(/No query restriction was specified/i)).toBeInTheDocument()
    } catch (error) {
      error.message = ` ${`Because it is a second try to search with empty params it must give a warning message\n\n ${error}`} `
      throw error
    }

    expect(asFragment()).toMatchSnapshot()
  })

  test('double search: show an overflow warning message and then show a message about empty results "noHits". English. 3A', async () => {
    useStore.mockReturnValue({
      browserConfig: { proxyPrefixPath: { uri: '/student/kurser' } },
      language: 'en',
      languageIndex: 0,
    })

    koppsCourseSearch.mockReturnValue(Promise.resolve(TEST_API_ANSWER_OVERFLOW))

    const { asFragment, rerender } = render(<SearchResultDisplay searchParameters={{ pattern: 'A' }} />)
    expect(screen.getByText(/searching/i)).toBeInTheDocument()
    expect(koppsCourseSearch).toHaveBeenCalledWith('en', '/student/kurser', { pattern: 'A' })

    await waitForElementToBeRemoved(() => screen.getByText(/searching/i))
    expect(screen.getByText(/Search results/i)).toBeInTheDocument()
    expect(screen.getByText(/There were too many results./i)).toBeInTheDocument()

    // rerender
    koppsCourseSearch.mockReturnValue(Promise.resolve(TEST_API_ANSWER_NO_HITS))

    act(() => rerender(<SearchResultDisplay searchParameters={{ pattern: 'CHOKLADKAKA' }} />))
    expect(koppsCourseSearch).toHaveBeenCalledWith('en', '/student/kurser', { pattern: 'CHOKLADKAKA' })
    await waitForElementToBeRemoved(() => screen.getByText(/searching/i))

    expect(screen.getByText(/Search results/i)).toBeInTheDocument()
    expect(screen.getByText(/Your search returned no results./i)).toBeInTheDocument()

    expect(asFragment()).toMatchSnapshot()
  })

  test('double search: show empty results message "noHits" and then show a rejected message of unknown error. English. 4A', async () => {
    useStore.mockReturnValue({
      browserConfig: { proxyPrefixPath: { uri: '/student/kurser' } },
      language: 'en',
      languageIndex: 0,
    })

    koppsCourseSearch.mockReturnValue(Promise.resolve(TEST_API_ANSWER_NO_HITS))

    const { asFragment, rerender } = render(<SearchResultDisplay searchParameters={{ pattern: 'CHOKLADKAKA' }} />)
    expect(screen.getByText(/searching/i)).toBeInTheDocument()
    expect(koppsCourseSearch).toHaveBeenCalledWith('en', '/student/kurser', { pattern: 'CHOKLADKAKA' })
    await waitForElementToBeRemoved(() => screen.getByText(/searching/i))

    expect(screen.getByText(/Search results/i)).toBeInTheDocument()
    expect(screen.getByText(/Your search returned no results./i)).toBeInTheDocument()

    expect(asFragment()).toMatchSnapshot()

    // rerender
    koppsCourseSearch.mockReturnValue(Promise.reject(TEST_API_ANSWER_UNKNOWN_ERROR))

    act(() => rerender(<SearchResultDisplay searchParameters={{ pattern: 'CHOKLADKAKA' }} />))
    expect(koppsCourseSearch).toHaveBeenCalledWith('en', '/student/kurser', { pattern: 'CHOKLADKAKA' })
    await waitForElementToBeRemoved(() => screen.getByText(/searching/i))

    expect(screen.getByText(/Search results/i)).toBeInTheDocument()
    expect(screen.getByText(/An unknown error occurred - failed to retrieve course data./i)).toBeInTheDocument()

    expect(asFragment()).toMatchSnapshot()
  })

  test('search:show a rejected message of unknown error. English. 5A', async () => {
    useStore.mockReturnValue({
      browserConfig: { proxyPrefixPath: { uri: '/student/kurser' } },
      language: 'en',
      languageIndex: 0,
    })

    // rerender
    koppsCourseSearch.mockReturnValue(Promise.reject(TEST_API_ANSWER_UNKNOWN_ERROR))

    const { asFragment } = render(<SearchResultDisplay searchParameters={{ pattern: 'CHOKLADKAKA' }} />)
    expect(screen.getByText(/searching/i)).toBeInTheDocument()

    expect(koppsCourseSearch).toHaveBeenCalledWith('en', '/student/kurser', { pattern: 'CHOKLADKAKA' })
    await waitForElementToBeRemoved(() => screen.getByText(/searching/i))

    expect(screen.getByText(/Search results/i)).toBeInTheDocument()
    expect(screen.getByText(/An unknown error occurred - failed to retrieve course data./i)).toBeInTheDocument()

    expect(asFragment()).toMatchSnapshot()
  })
})

describe.skip('Component <SearchResultDisplay> and resolved cases', () => {
  test('search by a text pattern to display at first a pending status message then the result of this search. English. 1B', async () => {
    useStore.mockReturnValue({
      browserConfig: { proxyPrefixPath: { uri: '/student/kurser' } },
      language: 'en',
      languageIndex: 0,
    })

    koppsCourseSearch.mockReturnValue(Promise.resolve(TEST_API_ANSWER_ALGEBRA))

    const { asFragment } = render(<SearchResultDisplay searchParameters={{ pattern: 'Algebra' }} />)

    expect(screen.getByText(/searching/i)).toBeInTheDocument()

    expect(koppsCourseSearch).toHaveBeenCalledWith('en', '/student/kurser', { pattern: 'Algebra' })

    await waitForElementToBeRemoved(() => screen.getByText(/searching/i))

    expect(screen.getByTestId('number-of-results')).toHaveTextContent('Your search returned 2 result(s).')

    const rows = screen.queryAllByRole('row')

    expect(rows).toHaveLength(3)
    expect(screen.getByText(/IX1303/i)).toBeInTheDocument()
    expect(screen.getByText(/SF1624/i)).toBeInTheDocument()

    expect(asFragment()).toMatchSnapshot()
  })

  test('double search: by deparment&pattern and then by education level. English. 2B', async () => {
    useStore.mockReturnValue({
      browserConfig: { proxyPrefixPath: { uri: '/student/kurser' } },
      language: 'en',
      languageIndex: 0,
    })

    koppsCourseSearch.mockReturnValue(Promise.resolve(TEST_API_ANSWER_ALGEBRA))

    const { asFragment, rerender } = render(
      <SearchResultDisplay searchParameters={{ pattern: 'Algebra', department: 'ABD' }} />
    )

    expect(screen.getByText(/searching/i)).toBeInTheDocument()

    expect(koppsCourseSearch).toHaveBeenCalledWith('en', '/student/kurser', { pattern: 'Algebra', department: 'ABD' })

    await waitForElementToBeRemoved(() => screen.getByText(/searching/i))

    expect(screen.getByTestId('number-of-results')).toHaveTextContent('Your search returned 2 result(s).')

    const rows = screen.queryAllByRole('row')

    expect(rows).toHaveLength(3)
    expect(screen.getByText(/IX1303/i)).toBeInTheDocument()
    expect(screen.getByText(/SF1624/i)).toBeInTheDocument()

    expect(asFragment()).toMatchSnapshot()

    // rerender, second search

    koppsCourseSearch.mockReturnValue(Promise.resolve(TEST_API_ANSWER_RESOLVED))

    rerender(
      <SearchResultDisplay
        searchParameters={{ eduLevel: ['0', '1', '2'], showOptions: ['onlyEnglish', 'showCancelled', 'onlyMHU'] }}
      />
    )

    expect(screen.getByText(/searching/i)).toBeInTheDocument()

    expect(koppsCourseSearch).toHaveBeenCalledWith('en', '/student/kurser', {
      eduLevel: ['0', '1', '2'],
      showOptions: ['onlyEnglish', 'showCancelled', 'onlyMHU'],
    })

    await waitForElementToBeRemoved(() => screen.getByText(/searching/i))

    expect(screen.getByTestId('number-of-results')).toHaveTextContent('Your search returned 2 result(s).')

    const newrows = screen.queryAllByRole('row')

    expect(newrows).toHaveLength(3)
    expect(screen.getByText(/AF2402/i)).toBeInTheDocument()
    expect(screen.getByText(/AH2905/i)).toBeInTheDocument()

    expect(asFragment()).toMatchSnapshot()
  })
})
