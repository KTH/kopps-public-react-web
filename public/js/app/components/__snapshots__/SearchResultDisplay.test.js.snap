// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component <SearchResultDisplay> and its warning messages double search: by a text pattern and rerender it again with the empty parameter to get a warning message noQueryProvided. English. 2A 1`] = `
<DocumentFragment>
  <h2
    id="results-heading"
  >
    Search results
  </h2>
  <p>
    <i>
      No query restriction was specified
    </i>
  </p>
</DocumentFragment>
`;

exports[`Component <SearchResultDisplay> and its warning messages double search: show an overflow warning message and then show a message about empty results "noHits". English. 3A 1`] = `
<DocumentFragment>
  <h2
    id="results-heading"
  >
    Search results
  </h2>
  <p>
    <i>
      Your search returned no results.
    </i>
  </p>
</DocumentFragment>
`;

exports[`Component <SearchResultDisplay> and its warning messages double search: show empty results message "noHits" and then show a rejected message of unknown error. English. 4A 1`] = `
<DocumentFragment>
  <h2
    id="results-heading"
  >
    Search results
  </h2>
  <p>
    <i>
      Your search returned no results.
    </i>
  </p>
</DocumentFragment>
`;

exports[`Component <SearchResultDisplay> and its warning messages double search: show empty results message "noHits" and then show a rejected message of unknown error. English. 4A 2`] = `
<DocumentFragment>
  <h2
    id="results-heading"
  >
    Search results
  </h2>
  <p>
    <i>
      An unknown error occurred - failed to retrieve course data.
    </i>
  </p>
</DocumentFragment>
`;

exports[`Component <SearchResultDisplay> and its warning messages double search: twice render empty parameters, first time without asking kopps, second time show a warning message. English. 1A 1`] = `<DocumentFragment />`;

exports[`Component <SearchResultDisplay> and its warning messages double search: twice render empty parameters, first time without asking kopps, second time show a warning message. English. 1A 2`] = `
<DocumentFragment>
  <h2
    id="results-heading"
  >
    Search results
  </h2>
  <p>
    <i>
      No query restriction was specified
    </i>
  </p>
</DocumentFragment>
`;

exports[`Component <SearchResultDisplay> and its warning messages search:show a rejected message of unknown error. English. 5A 1`] = `
<DocumentFragment>
  <h2
    id="results-heading"
  >
    Search results
  </h2>
  <p>
    <i>
      An unknown error occurred - failed to retrieve course data.
    </i>
  </p>
</DocumentFragment>
`;

exports[`Component <SearchResultDisplay> and resolved cases double search: by deparment&pattern and then by education level. English. 2B 1`] = `
<DocumentFragment>
  <h2
    id="results-heading"
  >
    Search results
  </h2>
  <div
    class="row"
  >
    <div
      class="col"
    >
      <div>
        <p
          class="typescript-component"
        >
          I'M A TYPESCRIPT COMPONENT
        </p>
        <article
          class="article article--no-start-margin "
        >
          <p
            data-testid="number-of-results"
          >
            Your search returned 
            <b>
              2
            </b>
             result(s).
          </p>
          <table
            class="table table-responsive"
          >
            <thead>
              <tr>
                <th
                  style="cursor: pointer; user-select: none;"
                >
                  Course code
                  <span
                    aria-hidden="true"
                    style="opacity: 0; margin-left: .25em;"
                  >
                    ▼
                  </span>
                </th>
                <th
                  style="cursor: pointer; user-select: none;"
                >
                  Course name
                  <span
                    aria-hidden="true"
                    style="opacity: 0; margin-left: .25em;"
                  >
                    ▼
                  </span>
                </th>
                <th
                  style="cursor: pointer; user-select: none;"
                >
                  Scope
                  <span
                    aria-hidden="true"
                    style="opacity: 0; margin-left: .25em;"
                  >
                    ▼
                  </span>
                </th>
                <th
                  style="cursor: pointer; user-select: none;"
                >
                  Educational level
                  <span
                    aria-hidden="true"
                    style="opacity: 0; margin-left: .25em;"
                  >
                    ▼
                  </span>
                </th>
              </tr>
            </thead>
            <tbody>
              <tr>
                <td>
                  <a
                    href="/student/kurser/kurs/IX1303?l=en"
                  >
                    IX1303
                  </a>
                </td>
                <td>
                  <a
                    href="/student/kurser/kurs/IX1303?l=en"
                  >
                    Algebra och geometri
                  </a>
                </td>
                <td>
                  7.5 hp
                </td>
                <td>
                  First cycle
                </td>
              </tr>
              <tr>
                <td>
                  <a
                    href="/student/kurser/kurs/SF1624?l=en"
                  >
                    SF1624
                  </a>
                </td>
                <td>
                  <a
                    href="/student/kurser/kurs/SF1624?l=en"
                  >
                    Algebra och geometri
                  </a>
                </td>
                <td>
                  7.5 hp
                </td>
                <td>
                  First cycle
                </td>
              </tr>
            </tbody>
          </table>
        </article>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Component <SearchResultDisplay> and resolved cases double search: by deparment&pattern and then by education level. English. 2B 2`] = `
<DocumentFragment>
  <h2
    id="results-heading"
  >
    Search results
  </h2>
  <div
    class="row"
  >
    <div
      class="col"
    >
      <div>
        <p
          class="typescript-component"
        >
          I'M A TYPESCRIPT COMPONENT
        </p>
        <article
          class="article article--no-start-margin "
        >
          <p
            data-testid="number-of-results"
          >
            Your search returned 
            <b>
              2
            </b>
             result(s).
          </p>
          <table
            class="table table-responsive"
          >
            <thead>
              <tr>
                <th
                  style="cursor: pointer; user-select: none;"
                >
                  Course code
                  <span
                    aria-hidden="true"
                    style="opacity: 0; margin-left: .25em;"
                  >
                    ▼
                  </span>
                </th>
                <th
                  style="cursor: pointer; user-select: none;"
                >
                  Course name
                  <span
                    aria-hidden="true"
                    style="opacity: 0; margin-left: .25em;"
                  >
                    ▼
                  </span>
                </th>
                <th
                  style="cursor: pointer; user-select: none;"
                >
                  Scope
                  <span
                    aria-hidden="true"
                    style="opacity: 0; margin-left: .25em;"
                  >
                    ▼
                  </span>
                </th>
                <th
                  style="cursor: pointer; user-select: none;"
                >
                  Educational level
                  <span
                    aria-hidden="true"
                    style="opacity: 0; margin-left: .25em;"
                  >
                    ▼
                  </span>
                </th>
                <th
                  style="cursor: pointer; user-select: none;"
                >
                  Periods
                  <span
                    aria-hidden="true"
                    style="opacity: 0; margin-left: .25em;"
                  >
                    ▼
                  </span>
                </th>
              </tr>
            </thead>
            <tbody>
              <tr>
                <td>
                  <a
                    href="/student/kurser/kurs/AF2402?startterm=20212&l=en"
                  >
                    AF2402
                  </a>
                </td>
                <td>
                  <a
                    href="/student/kurser/kurs/AF2402?startterm=20212&l=en"
                  >
                    Acoustics and Fire
                  </a>
                </td>
                <td>
                  7.5 hp
                </td>
                <td>
                  First cycle
                </td>
                <td>
                  P1 Autumn 21 - P3 Spring 22
                </td>
              </tr>
              <tr>
                <td>
                  <a
                    href="/student/kurser/kurs/AH2905?startterm=20212&l=en"
                  >
                    AH2905
                  </a>
                </td>
                <td>
                  <a
                    href="/student/kurser/kurs/AH2905?startterm=20212&l=en"
                  >
                    Advanced Pavement Engineering Analysis and Design
                  </a>
                </td>
                <td>
                  7.5 hp
                </td>
                <td />
                <td>
                  P1 Autumn 21
                </td>
              </tr>
            </tbody>
          </table>
        </article>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Component <SearchResultDisplay> and resolved cases search by a text pattern to display at first a pending status message then the result of this search. English. 1B 1`] = `
<DocumentFragment>
  <h2
    id="results-heading"
  >
    Search results
  </h2>
  <div
    class="row"
  >
    <div
      class="col"
    >
      <div>
        <p
          class="typescript-component"
        >
          I'M A TYPESCRIPT COMPONENT
        </p>
        <article
          class="article article--no-start-margin "
        >
          <p
            data-testid="number-of-results"
          >
            Your search returned 
            <b>
              2
            </b>
             result(s).
          </p>
          <table
            class="table table-responsive"
          >
            <thead>
              <tr>
                <th
                  style="cursor: pointer; user-select: none;"
                >
                  Course code
                  <span
                    aria-hidden="true"
                    style="opacity: 0; margin-left: .25em;"
                  >
                    ▼
                  </span>
                </th>
                <th
                  style="cursor: pointer; user-select: none;"
                >
                  Course name
                  <span
                    aria-hidden="true"
                    style="opacity: 0; margin-left: .25em;"
                  >
                    ▼
                  </span>
                </th>
                <th
                  style="cursor: pointer; user-select: none;"
                >
                  Scope
                  <span
                    aria-hidden="true"
                    style="opacity: 0; margin-left: .25em;"
                  >
                    ▼
                  </span>
                </th>
                <th
                  style="cursor: pointer; user-select: none;"
                >
                  Educational level
                  <span
                    aria-hidden="true"
                    style="opacity: 0; margin-left: .25em;"
                  >
                    ▼
                  </span>
                </th>
              </tr>
            </thead>
            <tbody>
              <tr>
                <td>
                  <a
                    href="/student/kurser/kurs/IX1303?l=en"
                  >
                    IX1303
                  </a>
                </td>
                <td>
                  <a
                    href="/student/kurser/kurs/IX1303?l=en"
                  >
                    Algebra och geometri
                  </a>
                </td>
                <td>
                  7.5 hp
                </td>
                <td>
                  First cycle
                </td>
              </tr>
              <tr>
                <td>
                  <a
                    href="/student/kurser/kurs/SF1624?l=en"
                  >
                    SF1624
                  </a>
                </td>
                <td>
                  <a
                    href="/student/kurser/kurs/SF1624?l=en"
                  >
                    Algebra och geometri
                  </a>
                </td>
                <td>
                  7.5 hp
                </td>
                <td>
                  First cycle
                </td>
              </tr>
            </tbody>
          </table>
        </article>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
